### PR TITLE
JSAB2-88 Persist user authentication

### PR DESCRIPTION
**Pull request by HONGTAO-T**

### CODE REVIEW

**Review by GuJialu — APPROVED** [review]

**Comment by HONGTAO-T** [comment]
[0001-JSAB2-88-test-persist-user-authentication.patch.txt](https://github.com/green-fox-academy/jsa-bloodstone-app/files/4012279/0001-JSAB2-88-test-persist-user-authentication.patch.txt)
In order to test the persist-user-authentication, please apply this patch to the branch, there will be a logout button that can change the 'isLoggedIn' state to false.

apply patch command:
```
git apply path/to/0001-JSAB2-88-test-persist-user-authentication.patch.txt
```

File: frontend/src/AuthLoadingScreen/index.jsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+} from 'react-native';
+import { useNavigation } from 'react-navigation-hooks';
+
+function AuthLoadingScreen() {
+  const navigation = useNavigation();
+  const isLoggedIn = useSelector((state) => state.auth.isLoggedIn);
+  if (isLoggedIn) {
+    return navigation.navigate('Home');
+  }
+  if (!isLoggedIn) {
```
Comment (kyya) [inline frontend/src/AuthLoadingScreen/index.jsx]: why you are not using else statement here?

File: frontend/src/Navigation/index.jsx [hunk 2]
```diff
@@ -7,6 +7,7 @@ import Settings from '../Settings';
 import colors from '../common/colors';
 import Game from '../Game';
 import Login from '../Login';
+import AuthLoadingScreen from '../AuthLoadingScreen';
```
Comment (kyya) [inline frontend/src/Navigation/index.jsx:13]: we are going to use an animation for loading, here are some [demos](https://dribbble.com/search/loading).

**Review by kyya — APPROVED** [review]

**Review by kyya — APPROVED** [review]

**Review by HanqingZ — APPROVED** [review]

File: frontend/src/store.js [hunk 3]
```diff
@@ -16,4 +19,14 @@ const rootReducer = combineReducers({
...
 });
 
-export default createStore(rootReducer, applyMiddleware(thunk));
+const persistConfig = {
+  key: 'root',
+  storage,
+  whitelist: ['auth'],
+};
+
+const pReducer = persistReducer(persistConfig, rootReducer);
```
Comment (evasimonyi) [inline frontend/src/store.js]: please rename pReducer into something more readable/understandable, like `persistedReducer` as it is suggested in the official doc: [https://github.com/rt2zz/redux-persist](url)

File: frontend/src/AuthLoadingScreen/index.jsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+function AuthLoadingScreen() {
+  const navigation = useNavigation();
+  const isLoggedIn = useSelector((state) => state.auth.isLoggedIn);
+  setTimeout(() => {
+    if (isLoggedIn) {
+      return navigation.navigate('Home');
+    }
+    return navigation.navigate('Auth');
+  }, 8000);
```
Comment (evasimonyi) [inline frontend/src/AuthLoadingScreen/index.jsx]: if I understand correctly, when the users open the app, they have to wait 8 seconds, you check if they are logged in, and after 8 seconds have passed, they are redirected? this is not a great user experience, also 8 seconds seems like a magic number, why 8 seconds exactly? 

but my main question is why did you even use `setTimeout`, when the `PersistGate`'s goal is exactly this: to wait for the needed data from the state. I might have just missed something, thanks for the answer!

**Review by HanqingZ — APPROVED** [review]

**Review by kyya — APPROVED** [review]

**Review by GuJialu — APPROVED** [review]

**Review by HanqingZ — APPROVED** [review]

**Review by evasimonyi — APPROVED** [review]

**Pull request merged**